instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarDetails.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.javaagent.runtimemetrics.java8;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.toMap;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.net.URL;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Properties;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarInputStream;
+import java.util.jar.Manifest;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+
+/**
+ * For a given URL representing a Jar directly on the file system or embedded within another
+ * archive, this class provides methods which expose useful information about it.
+ */
+class JarDetails {
```

Comment:
I find the arrangement of all this code into a class with inheritance and where the various extraction bits are distributed across construction and accessors to be harder to reason about than the JarAnalyzerUtil approach, where each extraction task is implemented as a pure function without side affects. 

Is there any way to have the best of both worlds? Simple code arrangement and good performance?

Reply:
Some of the things could be extracted to a static method, but many of the significant optimizations were unlocked in the overloaded methods. Do you have anything specific in mind?